hlib/python/hlib/frontend/relay_parser.py

```diff
@@ -0,0 +1,696 @@
+            call_env = (call_env[2])[call_var]
+            _var = call_env[0]
+            _dict = call_env[1]
+            _env = call_env[2]
+            _size = call_env[3]
+            new_params = []
+            for arg in _var:
+                if arg in params:
+                    new_params.append(arg)
+            _func = gen_func(new_params, _var, _dict, _env)
+            return _func(*call_args)
+        elif(call_type == Call):
+            _func = bind_env[call_var][1]
+            _args = bind_env[call_var][2]
+            _kwargs = bind_env[call_var][3]
+            _args = list(_args)
+            for i in range(len(_args)):
+                item = _args[i]
+                if isinstance(item, str):
+                    if type_dict[item] == Call:
+                        inner_func = bind_env[item][1]
+                        inner_args = bind_env[item][2]
+                        inner_kwargs = bind_env[item][3]
+                        _args[i] = inner_func(*inner_args, **inner_kwargs)
+            _args = tuple(_args)
+            arg_list = []
+            for _var in _args:
+                if _var in params:
+                    arg_list.append(_var)
+                else:
+                    arg_list.append(_var)
+            return _func(*arg_list, **_kwargs)
+    if ntype == Tuple:
+        var = arg[0][0]
+        env = arg[2]
+        tup = env[var][1]
+        dict_type = env[var][2]
+        tup_env = env[var][3]
+        return tuple_extract(tup, dict_type, tup_env)
+    if ntype == Var:
+        name = arg[0][0]
+        return (arg[2])[name]
+    else:
+        print("Type not implemented yet")
+
+
+def isInParams(var, params):
+    if(not isinstance(var, hcl.tensor.Tensor)):
+        return False
+
+    for par in params:
+        isShape = (var.shape == par.shape)
+        isName = (var.name == par.name)
+        isType = (var.dtype == par.dtype)
+        if(isShape and isName and isType):
+            return True
+    return False
+
+
+def get_type(ty, name):
+    if isinstance(ty, TensorType):
+        dtype = ty.dtype
+        size = []
+        for i in ty.shape:
+            size.append(i.value)
+        return hcl.placeholder(tuple(size), name, dtype)
+    elif isinstance(ty, TupleType):
+        t_vars = []
+        for i in range(len(ty.fields)):
+            var_name = name + "_" + str(i)
+            t_vars.append(get_type(ty.fields[i], var_name))
+        return tuple(t_vars)
+    else:
+        pass
+
+# get_Item
+
+
+def get_item(env):
+    tup_type = env[1]
+    if tup_type == Var:
+        tup = list(env[2])
+        index = env[3]
+        item = tup[index]
+        if(isinstance(item, tuple)):
+            name = env[0]
+        else:
+            name = item.name
+        inst_type = {name: Var}
+        inst_env = {name: item}
+    if tup_type == Call:
+        name = env[0]
+        tup = env[2]
+        index = env[3]
+    return item, name, inst_type, inst_env, inst_var
```

Comment:
Does that mean ``inst_env`` and ``inst_var`` could be NULL?

Reply:
Yes